projects/plugins/jetpack/src/class-deprecate.php

```diff
@@ -21,12 +22,34 @@ class Deprecate {
 	 */
 	private static $instance;
 
+	/**
+	 * An array of notices to display.
+	 *
+	 * @var array
+	 */
+	private $notices = array();
+
 	/**
 	 * Initialize the class.
 	 */
 	private function __construct() {
+		// phpcs:disable Squiz.PHP.CommentedOutCode.Found
+		// Example $this->notices content. Note that the key must end with '-deprecate-feature' to display in My Jetpack - see packages/my-jetpack/_inc/hooks/use-notification-watcher/use-deprecate-feature-notice.ts:
+		// array(
+		// 'my-admin-deprecate-feature' => array(
+		// 'title'     => __( "Retired feature: Jetpack's XYZ Feature", 'jetpack' ),
+		// 'message'   => __( 'This feature is being retired and will be removed effective November, 2024. Please use the Classic Theme Helper plugin instead.', 'jetpack' ),
+		// 'link'      => array(
+		// 'label' => __( 'Learn more', 'jetpack' ),
+		// 'url'   => 'jetpack-support-xyz',
+		// ),
+		// 'show'      => $this->show_feature_notice( 'my-admin-deprecate-feature' ),
```

Comment:
I was also thinking of an alternate approach that could facilitate consumers of this class and also introduce some validation for the newly added notices:

The consumers can add their notices in the private `$notices` property. No need to provide translated title and message. The notice key doesn't need to end with `-deprecate-feature` (we'll take care of that during validation - see `set_notices` below). If `show` is set to `false` the notice won't get displayed but can still be present in the private class property (this way we can keep our example set here instead of commented-out code).

```
private $notices = array(
	'my-admin' => array(
		'title'     => "Retired feature: Jetpack's XYZ Feature",
		'message'   => 'This feature is being retired and will be removed effective November, 2024. Please use the Classic Theme Helper plugin instead.',
		'link'      => array(
			'url'   => 'jetpack-support-xyz',
		),
		'show'      => true,
	),
);
```

We validate and set up notices in the constructor via `set_notices`:
```
private function __construct() {
    $this->set_notices();
	
    if ( $this->has_notices() ) {
		add_action( 'load-index.php', array( $this, 'render_admin_notices' ) );
		add_action( 'admin_enqueue_scripts', array( $this, 'enqueue_admin_scripts' ) );
		add_filter( 'my_jetpack_red_bubble_notification_slugs', array( $this, 'add_my_jetpack_red_bubbles' ) 
    );
   }
}
```

`set_notices` will validate the provided ones, remove the ones that don't have required fields eg `title`, `message` or `link`. Also append the `-deprecate-feature` if not present and set a default value for the link label: `Learn more`
```
private function set_notices() {
	$notices            = array();
	$required_id_suffix = '-deprecate-feature';

	foreach ( $this->notices as $id => $notice ) {
		if ( isset( $notice['show'] ) && false === $notice['show'] ) {
			continue;
		}
		if ( empty( $notice['title'] ) || empty( $notice['message'] ) || empty( $notice['link']['url'] ) ) {
			continue;
		}
		$notice['title']   = __( $notice['title'], 'jetpack' );
		$notice['message'] = __( $notice['message'], 'jetpack' );

		if ( empty( $notice['link']['label'] ) ) {
			$notice['link']['label'] = 'Learn more';
		}
		$notice['link']['label'] = __( $notice['link']['label'], 'jetpack' );

		if ( strpos( $id, $required_id_suffix ) === false ) {
			$id .= $required_id_suffix;
		}

		$notices[ $id ] = $notice;
	}

	$this->notices = $notices;
}
```

Reply:
> I was wondering if we actually need show as a notice property. It looks like it repeats the show_feature_notice check but it's only used by DeprecationNotice. Therefore, if show is set to false, the notice will still display in eg MyJetpack, right?

Yes it is just for `DeprecationNotice`. But removing it would mean adding in logic back in modified logic in `jetpack-notices/index.jsx` to check for the existence of the cookie, which seems unnecessary given it can be added as a single property in this current way. Curious to know what you think though - if the separation of logic is better in this case?

> I was also thinking of an alternate approach that could facilitate consumers of this class and also introduce some validation for the newly added notices

I love this! I'll take a look at implementing it - I had one question though:

> The consumers can add their notices in the private $notices property.

It looks like the best practice is to do what you've suggested if none of the variables will change, but otherwise initialize it within `__construct` if the content is dynamic. So if the 'show' property remained, then keeping it in __construct makes sense. Is that your understanding as well?

Reply:
Oh! I see that now, thanks for clarifying!

In that case, I totally see how it makes sense to remove the show property - but keep it for the example case.

I've applied these changes with a slight modification - since it isn't recommended to pass in a variable for translation with no other text, I could work around that by having partial strings, such as `$notice['title'] = sprintf( __( 'Retired feature: %s', 'jetpack'), $notice['title']);`, but it get's tricky for the label and message. So we're better off translating the strings at source. And if we're doing that, it's considered dynamic content anyway, so we still can't add the $notices array just when it is declared. I've kept it in the `__construct` function.
Changes in 0099086faf63510cff7490c606f7f8b044cb79cd

Reply:
That makes sense, thank you!